File: internal/pool/backend.go
// Package pool manages backend servers and provides algorithms for selecting
// which backend should handle each incoming request.
//
// Architecture:
//   - Backend: represents a single upstream server with its own reverse proxy,
//     health status, and connection tracking. All mutable state uses atomic
//     operations for lock-free concurrent access.
//   - BackendPool: interface for backend selection algorithms. Phase 2 provides
//     RoundRobinPool; Phase 3 will add LeastConnections, Weighted, IPHash, etc.
//
// Concurrency design:
//   Each incoming HTTP request runs in its own goroutine (created by http.Server).
//   All these goroutines call pool.Next() concurrently to get a backend. The pool
//   must be safe for concurrent access without introducing contention.
//
//   We achieve this through:
//   - atomic.Uint64 for the round-robin counter (lock-free, ~5ns per operation)
//   - atomic.Bool for backend alive status (lock-free reads on every request)
//   - atomic.Int64 for active connection counts (lock-free increment/decrement)
//   - sync.RWMutex on the backend slice (only write-locked when adding/removing backends)
//
// Why a separate package from proxy?
//   Separation of concerns. The proxy package handles HTTP forwarding (Director,
//   ModifyResponse, ErrorHandler). The pool package handles backend lifecycle and
//   selection. This makes each package independently testable and allows swapping
//   selection algorithms without touching HTTP logic.
package pool

import (
	"fmt"
	"net"
	"net/http"
	"net/http/httputil"
	"net/url"
	"sync/atomic"
	"time"

	"go.uber.org/zap"
)

// Backend represents a single upstream server.
//
// Each backend has its own httputil.ReverseProxy instance with an independent
// connection pool (http.Transport). This is a deliberate design choice:
//
// Alternative: Share a single Transport across all backends.
// Problem: A slow backend (e.g., experiencing GC pauses) would exhaust shared
// connections, blocking requests to healthy backends. Per-backend transports
// provide fault isolation — a slow backend only exhausts its own pool.
//
// Memory cost: ~1KB per backend for the struct + Transport overhead.
// For typical deployments (2-20 backends), this is negligible.
type Backend struct {
	// url is the parsed URL of the backend server.
	url *url.URL

	// name is a human-readable identifier for logging and dashboards.
	name string

	// weight determines the relative traffic share in weighted algorithms.
	// Higher weight = more traffic. Ignored by simple round-robin.
	// Default: 1. Range: 1-100.
	weight int

	// alive indicates whether the backend is currently healthy.
	// Updated by health checks (Phase 4). Starts as true.
	//
	// Uses atomic.Bool for lock-free reads. Every request goroutine reads
	// this to decide whether to skip this backend. A mutex would create
	// unnecessary contention under high load.
	alive atomic.Bool

	// activeConnections tracks the number of in-flight requests.
	// Incremented when a request starts, decremented when it completes.
	//
	// Used by:
	//   - LeastConnections algorithm (Phase 3) for backend selection
	//   - MaxConnections enforcement (connection limiting)
	//   - Metrics/dashboards for monitoring
	//
	// Uses atomic.Int64 for lock-free increment/decrement.
	activeConnections atomic.Int64

	// maxConnections is the maximum number of concurrent connections allowed.
	// 0 means unlimited. When the limit is reached, the backend is temporarily
	// skipped during selection (treated as "at capacity" rather than "unhealthy").
	maxConnections int

	// proxy is the httputil.ReverseProxy instance for this backend.
	// Each backend gets its own proxy with its own Transport (connection pool).
	proxy *httputil.ReverseProxy

	// logger is the structured logger for this backend's operations.
	logger *zap.Logger
}

// NewBackend creates a new Backend with a configured reverse proxy.
//
// Parameters:
//   - rawURL: Backend server URL (e.g., "http://localhost:9001")
//   - name: Human-readable name for logging (e.g., "backend-1")
//   - weight: Traffic weight for weighted algorithms (1-100, default 1)
//   - maxConns: Max concurrent connections (0 = unlimited)
//   - logger: Structured logger
//
// Each backend gets its own httputil.ReverseProxy and http.Transport.
// This provides connection pool isolation between backends.
func NewBackend(rawURL string, name string, weight int, maxConns int, logger *zap.Logger) (*Backend, error) {
	parsedURL, err := url.Parse(rawURL)
	if err != nil {
		return nil, fmt.Errorf("parsing backend URL %q: %w", rawURL, err)
	}

	if weight <= 0 {
		weight = 1
	}

	b := &Backend{
		url:            parsedURL,
		name:           name,
		weight:         weight,
		maxConnections: maxConns,
		logger:         logger,
	}

	// Start as alive. Health checks (Phase 4) will update this.
	b.alive.Store(true)

	// Create the reverse proxy for this backend.
	b.proxy = &httputil.ReverseProxy{
		Director:       b.director,
		ModifyResponse: b.modifyResponse,
		ErrorHandler:   b.errorHandler,
		Transport:      newTransport(),
	}

	return b, nil
}

// URL returns the backend's URL string.
func (b *Backend) URL() string {
	return b.url.String()
}

// Name returns the backend's human-readable name.
func (b *Backend) Name() string {
	return b.name
}

// Weight returns the backend's traffic weight.
func (b *Backend) Weight() int {
	return b.weight
}

// IsAlive returns whether the backend is currently healthy.
//
// This is called on every request by the pool's Next() method.
// Uses atomic.Bool.Load() — lock-free, ~1ns.
func (b *Backend) IsAlive() bool {
	return b.alive.Load()
}

// SetAlive updates the backend's health status.
//
// Called by health checks (Phase 4) and circuit breakers (Phase 7).
// Uses atomic.Bool.Store() — lock-free, ~1ns.
func (b *Backend) SetAlive(alive bool) {
	b.alive.Store(alive)
	b.logger.Info("backend status changed",
		zap.String("backend", b.name),
		zap.Bool("alive", alive),
	)
}

// ActiveConnections returns the current number of in-flight requests.
func (b *Backend) ActiveConnections() int64 {
	return b.activeConnections.Load()
}

// IsAtCapacity returns true if the backend has reached its connection limit.
// Returns false if maxConnections is 0 (unlimited).
func (b *Backend) IsAtCapacity() bool {
	if b.maxConnections == 0 {
		return false
	}
	return b.activeConnections.Load() >= int64(b.maxConnections)
}

// ServeHTTP forwards the request to this backend.
//
// It tracks active connections: increments before forwarding, decrements
// after the response is complete (or on error). This ensures the count
// is always accurate, even if the backend crashes mid-response.
//
// The defer pattern guarantees decrement happens regardless of panics
// or errors in the proxy pipeline.
func (b *Backend) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	b.activeConnections.Add(1)
	defer b.activeConnections.Add(-1)

	b.proxy.ServeHTTP(w, r)
}

// director rewrites the request to target this backend.
//
// Identical logic to Phase 1, but now scoped per-backend.
// Each backend's Director targets its own URL.
func (b *Backend) director(req *http.Request) {
	req.URL.Scheme = b.url.Scheme
	req.URL.Host = b.url.Host
	req.Host = b.url.Host

	// Add X-Real-IP with the client's IP.
	if clientIP, _, err := net.SplitHostPort(req.RemoteAddr); err == nil {
		req.Header.Set("X-Real-IP", clientIP)
	}

	// Add X-Backend-Name so the client can see which backend handled the request.
	// Useful for debugging load balancing distribution.
	req.Header.Set("X-Backend-Name", b.name)
}

// modifyResponse adds proxy identification headers.
func (b *Backend) modifyResponse(resp *http.Response) error {
	resp.Header.Set("X-Proxy", "load-balancer")
	resp.Header.Set("X-Backend-Name", b.name)
	resp.Header.Add("Via", "1.1 load-balancer")
	return nil
}

// errorHandler handles connection-level errors (backend unreachable, timeout).
func (b *Backend) errorHandler(w http.ResponseWriter, r *http.Request, err error) {
	b.logger.Error("backend error",
		zap.String("backend", b.name),
		zap.String("url", b.url.String()),
		zap.String("method", r.Method),
		zap.String("path", r.URL.Path),
		zap.Error(err),
	)

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusBadGateway)
	fmt.Fprintf(w, `{"error": "bad gateway", "message": "backend %s unavailable", "backend": "%s", "status": 502}`,
		b.name, b.name)
}

// newTransport creates an http.Transport configured for reverse proxy workloads.
//
// Each backend gets its own Transport for connection pool isolation.
// See Phase 1 documentation for tuning rationale.
func newTransport() *http.Transport {
	return &http.Transport{
		MaxIdleConns:        100,
		MaxIdleConnsPerHost: 10,
		IdleConnTimeout:     90 * time.Second,
		DialContext: (&net.Dialer{
			Timeout:   30 * time.Second,
			KeepAlive: 30 * time.Second,
		}).DialContext,
		TLSHandshakeTimeout:   10 * time.Second,
		ResponseHeaderTimeout: 30 * time.Second,
		ExpectContinueTimeout: 1 * time.Second,
		ForceAttemptHTTP2:      true,
	}
}

File: src/main/java/com/loadbalancer/pool/Backend.java
package com.loadbalancer.pool;

import com.loadbalancer.circuit.CircuitBreaker;
import com.loadbalancer.config.CircuitBreakerConfig;
import com.loadbalancer.proxy.CapturedRequest;
import com.loadbalancer.proxy.ProxyResult;
import com.sun.net.httpserver.HttpExchange;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Represents a single upstream backend server.
 *
 * <p>Equivalent to Go's Backend struct. Each backend has its own {@link HttpClient}
 * instance (equivalent to Go's per-backend http.Transport) for connection pool
 * isolation — a slow backend won't exhaust connections meant for healthy ones.
 *
 * <p><b>Concurrency design:</b>
 * <ul>
 *   <li>{@code alive} — {@link AtomicBoolean} for lock-free reads (every request checks this)</li>
 *   <li>{@code activeConnections} — {@link AtomicLong} for lock-free increment/decrement</li>
 *   <li>{@code circuitBreaker} — per-backend state machine with volatile state (~1ns read)</li>
 *   <li>{@code httpClient} — thread-safe by design (Java HttpClient is immutable after build)</li>
 * </ul>
 *
 * <p>Memory cost: ~1KB per backend for the struct + HttpClient overhead.
 * For typical deployments (2-20 backends), this is negligible.
 */
public class Backend {

    private static final Logger logger = LoggerFactory.getLogger(Backend.class);

    private final URI url;
    private final String name;
    private final int weight;
    private final int maxConnections;
    private final HttpClient httpClient;
    private final CircuitBreaker circuitBreaker;

    /**
     * Whether this backend is currently healthy.
     * Updated by health checks (future phase). Starts as true.
     * Uses AtomicBoolean for lock-free reads — every request goroutine reads this.
     */
    private final AtomicBoolean alive = new AtomicBoolean(true);

    /**
     * Number of in-flight requests to this backend.
     * Incremented when a request starts, decremented when it completes.
     * Uses AtomicLong for lock-free increment/decrement.
     */
    private final AtomicLong activeConnections = new AtomicLong(0);

    /**
     * Creates a new Backend with a dedicated HttpClient and circuit breaker.
     *
     * @param url            Backend URL (e.g., "http://localhost:9001")
     * @param name           Human-readable name for logging
     * @param weight         Traffic weight for weighted algorithms (1-100)
     * @param maxConnections Max concurrent connections (0 = unlimited)
     * @param cbConfig       Circuit breaker configuration
     */
    public Backend(String url, String name, int weight, int maxConnections,
                   CircuitBreakerConfig cbConfig) {
        this.url = URI.create(url);
        this.name = name;
        this.weight = Math.max(weight, 1);
        this.maxConnections = maxConnections;

        // Each backend gets its own HttpClient for connection pool isolation.
        // Equivalent to Go's per-backend http.Transport.
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(30))
                .followRedirects(HttpClient.Redirect.NEVER)
                .build();

        this.circuitBreaker = new CircuitBreaker(name, cbConfig);
    }

    /**
     * Creates a new Backend with default circuit breaker configuration.
     * Preserves backward compatibility with existing tests.
     */
    public Backend(String url, String name, int weight, int maxConnections) {
        this(url, name, weight, maxConnections,
                CircuitBreakerConfig.withDefaults(null, null, null));
    }

    // --- Getters ---

    public URI url() { return url; }
    public String name() { return name; }
    public int weight() { return weight; }

    /**
     * Returns whether the backend is currently healthy.
     * Lock-free atomic read — ~1ns.
     */
    public boolean isAlive() {
        return alive.get();
    }

    /**
     * Returns whether the backend is available for traffic.
     * Combines health check status (alive) with circuit breaker state.
     *
     * <p>Uses {@code canAcceptTraffic()} (read-only) instead of
     * {@code allowRequest()} to avoid consuming the HALF_OPEN probe permit
     * during pool iteration. The actual permit is acquired in
     * {@link #forwardRequest(CapturedRequest)}.
     *
     * <p>A backend is available if:
     * <ul>
     *   <li>It's alive (health checks passing)</li>
     *   <li>The circuit breaker can accept traffic (not OPEN, or timeout elapsed)</li>
     * </ul>
     */
    public boolean isAvailable() {
        return alive.get() && circuitBreaker.canAcceptTraffic();
    }

    /**
     * Returns the circuit breaker for this backend.
     * Exposed for testing and metrics.
     */
    public CircuitBreaker circuitBreaker() {
        return circuitBreaker;
    }

    /**
     * Updates the backend's health status.
     * Called by health checks and circuit breakers.
     */
    public void setAlive(boolean alive) {
        this.alive.set(alive);
        logger.info("Backend status changed: backend={}, alive={}", name, alive);
    }

    /**
     * Returns the current number of in-flight requests.
     */
    public long activeConnections() {
        return activeConnections.get();
    }

    /**
     * Returns true if the backend has reached its connection limit.
     * Returns false if maxConnections is 0 (unlimited).
     */
    public boolean isAtCapacity() {
        if (maxConnections == 0) return false;
        return activeConnections.get() >= maxConnections;
    }

    /**
     * Forwards an HTTP request to this backend and writes the response back.
     *
     * <p>Convenience method that captures the request, forwards it, and writes
     * the result. Used when retries are not needed (backward compatibility).
     *
     * @param exchange the incoming HTTP exchange to proxy
     */
    public void handleRequest(HttpExchange exchange) throws IOException {
        CapturedRequest captured = CapturedRequest.from(exchange);
        ProxyResult result = forwardRequest(captured);
        result.writeTo(exchange);
    }

    /**
     * Forwards a captured request to this backend and returns the result
     * WITHOUT writing to the client.
     *
     * <p>This is the core method that enables retries. The caller (ProxyHandler)
     * can inspect the result and decide whether to retry with a different backend
     * or commit the result to the client.
     *
     * <p>Circuit breaker integration: if the circuit is OPEN, returns immediately
     * without making a network call. On success/failure, records the outcome to
     * the circuit breaker for future decisions.
     *
     * <p>Active connections are tracked atomically: incremented before forwarding,
     * decremented after response completes (or on error). The try/finally pattern
     * guarantees the count is always accurate.
     *
     * @param request the captured request to forward
     * @return the result of the forwarding attempt (never null)
     */
    public ProxyResult forwardRequest(CapturedRequest request) {
        // Circuit breaker gate — short-circuit without network call
        if (!circuitBreaker.allowRequest()) {
            return ProxyResult.circuitOpen(name);
        }

        activeConnections.incrementAndGet();
        try {
            // Build the target URL
            String targetUrl = buildTargetUrl(request.requestURI());

            // Build the HttpRequest
            HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                    .uri(URI.create(targetUrl))
                    .timeout(Duration.ofSeconds(30));

            // Set method + body
            if (request.body() != null) {
                reqBuilder.method(request.method(),
                        HttpRequest.BodyPublishers.ofByteArray(request.body()));
            } else if (request.bodyStream() != null) {
                reqBuilder.method(request.method(),
                        HttpRequest.BodyPublishers.ofInputStream(() -> request.bodyStream()));
            } else {
                reqBuilder.method(request.method(),
                        HttpRequest.BodyPublishers.noBody());
            }

            // Forward filtered client headers
            forwardRequestHeaders(reqBuilder, request.headers());

            // Add proxy identification headers
            addProxyHeaders(reqBuilder, request);

            // Send request to backend
            HttpResponse<java.io.InputStream> response = httpClient.send(
                    reqBuilder.build(),
                    HttpResponse.BodyHandlers.ofInputStream()
            );

            ProxyResult result = new ProxyResult(
                    response.statusCode(),
                    response.headers().map(),
                    response.body(),
                    name,
                    null
            );

            // Record outcome to circuit breaker
            if (result.isRetriable()) {
                circuitBreaker.recordFailure();
            } else {
                circuitBreaker.recordSuccess();
            }

            return result;

        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            circuitBreaker.recordFailure();
            return ProxyResult.connectionError(name, e);
        } catch (Exception e) {
            logger.error("Backend error: backend={}, url={}, error={}", name, url, e.getMessage());
            circuitBreaker.recordFailure();
            return ProxyResult.connectionError(name, e);
        } finally {
            activeConnections.decrementAndGet();
        }
    }

    // --- Private helpers ---

    /**
     * Builds the target URL by combining backend URL with the request path/query.
     */
    private String buildTargetUrl(URI requestURI) {
        String targetUrl = url.toString();
        if (targetUrl.endsWith("/")) {
            targetUrl = targetUrl.substring(0, targetUrl.length() - 1);
        }
        String requestPath = requestURI.getRawPath();
        targetUrl += requestPath;
        String query = requestURI.getRawQuery();
        if (query != null && !query.isEmpty()) {
            targetUrl += "?" + query;
        }
        return targetUrl;
    }

    /**
     * Forwards filtered client headers to the backend request.
     * Hop-by-hop and restricted headers are skipped.
     */
    private void forwardRequestHeaders(HttpRequest.Builder reqBuilder,
                                        Map<String, List<String>> headers) {
        if (headers == null) return;
        headers.forEach((key, values) -> {
            for (String value : values) {
                reqBuilder.header(key, value);
            }
        });
    }

    /**
     * Adds proxy identification headers (X-Real-IP, X-Forwarded-For, X-Backend-Name).
     */
    private void addProxyHeaders(HttpRequest.Builder reqBuilder, CapturedRequest request) {
        String clientIP = extractClientIP(request);
        if (clientIP != null) {
            reqBuilder.header("X-Real-IP", clientIP);
            // Check for existing X-Forwarded-For in the captured headers
            List<String> xffValues = request.headers().get("X-Forwarded-For");
            if (xffValues == null) {
                xffValues = request.headers().get("x-forwarded-for");
            }
            String existingXFF = (xffValues != null && !xffValues.isEmpty())
                    ? xffValues.get(0) : null;
            if (existingXFF != null && !existingXFF.isEmpty()) {
                reqBuilder.header("X-Forwarded-For", existingXFF + ", " + clientIP);
            } else {
                reqBuilder.header("X-Forwarded-For", clientIP);
            }
        }
        reqBuilder.header("X-Backend-Name", name);
    }

    /**
     * Extracts the client IP from the captured request's remote address.
     */
    private String extractClientIP(CapturedRequest request) {
        if (request.remoteAddress() != null && request.remoteAddress().getAddress() != null) {
            return request.remoteAddress().getAddress().getHostAddress();
        }
        return null;
    }

    /**
     * Sends a JSON error response to the client.
     * Equivalent to Go's errorHandler function.
     */
    private void sendErrorResponse(HttpExchange exchange, int statusCode, String message)
            throws IOException {
        String body = String.format(
                "{\"error\":\"bad gateway\",\"message\":\"%s\",\"backend\":\"%s\",\"status\":%d}",
                message, name, statusCode);
        byte[] bytes = body.getBytes();
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(statusCode, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    @Override
    public String toString() {
        return String.format("Backend{name='%s', url='%s', weight=%d, alive=%s, circuit=%s, activeConns=%d}",
                name, url, weight, alive.get(), circuitBreaker.state(), activeConnections.get());
    }
}
